tracee-rules/engine/engine.go

```diff
@@ -59,29 +67,46 @@ func NewEngine(sigs []types.Signature, sources EventSources, output chan types.F
 			continue
 		}
 	}
-	return &engine
+	return &engine, nil
+}
+
+// signatureStartFunc is the signature handling business logics.
+func signatureStartFunc(signature types.Signature, c chan types.Event) {
+	for e := range c {
+		if err := signature.OnEvent(e); err != nil {
+			meta, _ := signature.GetMetadata()
+			log.Printf("error handling event by signature %s: %v", meta.Name, err)
+		}
+	}
 }
 
 // Start starts processing events and detecting signatures
 // it runs continuously until stopped by the done channel
 // once done, it cleans all internal resources, which means the engine is not reusable
 // note that the input and output channels are created by the consumer and therefore are not closed
 func (engine *Engine) Start(done chan bool) {
+	defer engine.unloadAllSignatures()
+	engine.signaturesMutex.RLock()
 	for s, c := range engine.signatures {
-		defer close(c)
-		go func(s types.Signature, c chan types.Event) {
-			for e := range c {
-				err := s.OnEvent(e)
-				if err != nil {
-					meta, _ := s.GetMetadata()
-					log.Printf("error handling event by signature %s: %v", meta.Name, err)
-				}
-			}
-		}(s, c)
+		go signatureStartFunc(s, c)
 	}
+	engine.signaturesMutex.RUnlock()
 	engine.consumeSources(done)
 }
 
+func (engine *Engine) unloadAllSignatures() {
+	engine.signaturesMutex.Lock()
+	defer engine.signaturesMutex.Unlock()
+	for sig, c := range engine.signatures {
+		sig.Close()
+		close(c)
+		delete(engine.signatures, sig)
+	}
+	for selectedEvents, _ := range engine.signaturesIndex {
+		engine.signaturesIndex[selectedEvents] = nil
+	}
```

Comment:
this doesn't really clean the map as the keys will remain. you should copy L39 here